src/outputs/parthenon_hdf5.cpp

```diff
@@ -192,436 +330,652 @@ void PHDF5Output::genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm) {
+                local_count.data(), global_count.data(), pl_xfer);
+
+    // reset for collective output
     local_count[0] = num_blocks_local;
-    local_count[1] = nx3;
-    local_count[2] = nx2;
-    local_count[3] = nx1;
-    local_count[4] = 1;
-
-    global_count[0] = max_blocks_global;
-    global_count[1] = nx3;
-    global_count[2] = nx2;
-    global_count[3] = nx1;
-    global_count[4] = 1;
-
-    H5S const local_DSpace = H5S::FromHIDCheck(H5Screate_simple(5, local_count, NULL));
-    H5S const global_DSpace = H5S::FromHIDCheck(H5Screate_simple(5, global_count, NULL));
-
-    // while we could do this as n variables and load all variables for
-    // a block at one time, this is memory-expensive.  I think it is
-    // well worth the multiple iterations through the blocks to load up
-    // one variable at a time.  Besides most of the time will be spent
-    // writing the HDF5 file to disk anyway...
-    // If I'm wrong about this, we can always rewrite this later.
-    // Sriram
-
-    // this is a stupidly complicated multi-pass through the variable
-    // list, but again will revisit when the time comes to redo
-    for (auto &vwrite : ciX.vars) { // for each variable we write
-      const std::string vWriteName = vwrite->label();
-      hid_t vLocalSpace, vGlobalSpace;
-      H5S vLocalSpaceNew, vGlobalSpaceNew;
-      const hsize_t vlen = vwrite->GetDim(4);
-      local_count[4] = global_count[4] = vlen;
-
-      if (vlen == 1) {
-        vLocalSpace = local_DSpace;
-        vGlobalSpace = global_DSpace;
-      } else {
-        vLocalSpace = vLocalSpaceNew =
-            H5S::FromHIDCheck(H5Screate_simple(5, local_count, NULL));
-        vGlobalSpace = vGlobalSpaceNew =
-            H5S::FromHIDCheck(H5Screate_simple(5, global_count, NULL));
-      }
+  }
+
+  // -------------------------------------------------------------------------------- //
+  //   WRITING VARIABLES DATA                                                         //
+  // -------------------------------------------------------------------------------- //
+
+  // first we need to get list of variables, because sparse variables are only
+  // expanded on some blocks, we need to look at the list of variables on each block
+  // combine these into a global list of variables
+
+  auto get_MeshBlockDataIterator = [=](const std::shared_ptr<MeshBlock> pmb) {
+    if (restart_) {
+      return MeshBlockDataIterator<Real>(
+          pmb->meshblock_data.Get(),
+          {parthenon::Metadata::Independent, parthenon::Metadata::Restart}, true);
+    } else {
+      return MeshBlockDataIterator<Real>(pmb->meshblock_data.Get(),
+                                         output_params.variables);
+    }
+  };
+
+  // local list of unique vars
+  std::set<VarInfo> all_unique_vars;
+  for (auto &pmb : pm->block_list) {
+    auto ci = get_MeshBlockDataIterator(pmb);
+    for (auto &v : ci.vars) {
+      VarInfo vinfo(v);
+      all_unique_vars.insert(vinfo);
+    }
+  }
```

Comment:
So I'm not sure which is better, but an alternate approach to looping over meshblocks would be to query the state descriptor, which will be identical between meshblocks and contain relevant variable information.

There's a package called `parthenon::resolved_state`, which contains a unique list of all variables:
https://github.com/lanl/parthenon/blob/develop/src/interface/state_descriptor.cpp#L333

Reply:
This approach would also save you an allgather. But it would require parsing the variable labels kind of "by hand"

Reply:
So the issue here is that not all mesh blocks have the same variable fields. For example, we might have a sparse variable called "Sparse". Meshblock 0 has "Sparse_0", meshblock 1 has "Sparse_13" and "Sparse_2975", and meshblock 2 has "Sparse_0", "Sparse_13", and "Sparse_27". We now need to get a list of all the sparse IDs that "Sparse" has on any meshblock (globally). This is different from the resolved state that resolves name clashes, as far as I understand it. I don't think the StateDescriptor currently knows about the sparse IDs that may be allocated on any number of meshblocks, or does it? It just knows that there is a sparse variable called "Sparse", right? (At least, this will be true in #383, because there the sparse_id is removed from the Metadata.)

Reply:
How do you declare which sparse ids should exist at all? That has to be stored somewhere. I would think the `StateDescriptor` would know?

Reply:
Currently, the `SparseVariable` holds the allocated sparse ids on a per meshblock basis (`MeshblockData` has a list of `SparseVariable`s). So the `StateDescriptor` doesn't currently know which sparse ids exist. It only knows the names of the sparse variables. But you may be right that the `StateDescriptor` should know all the sparse ids that are allocated. However, we will still need a global allgather to exchange that information between all the ranks, unless we want to make allocating a new sparse id a global operation.

In fact, maybe there is a case to be made that all the sparse ids that can be used during the simulation have to be specified when the sparse variable is created. Then the list of possible sparse ids (per sparse variable) would live on the entire mesh (all ranks) from the very beginning, which will simply some things. We would still allocate particular sparse ids on individual blocks as needed, but at least the pool of ids is known globally ahead of time.

However, if we go this route, this would be implemented in #383 and then we could change this logic in the output code in that PR, so I think we should still merge what's in this PR now, even though it might be short lived...